lumigator/python/mzai/backend/backend/services/datasets.py

```diff
@@ -168,20 +167,33 @@ def get_dataset_s3_path(self, dataset_id: UUID) -> str:
         return dataset_path
 
     def delete_dataset(self, dataset_id: UUID) -> None:
+        """When the dataset exists, attempts to delete it from both an S3 bucket and the DB.
+
+        Raises exceptions it encounters dealing with S3, except when the file is not found in S3
+        (in this case it deletes the orphaned record from the DB).
+
+        This operation is idempotent, calling it with a record that never existed, or that has
+        already been deleted, will not raise an error.
+        """
         record = self._get_dataset_record(dataset_id)
+        # Early return if the record does not exist (for idempotency).
+        if record is None:
+            return None
 
         try:
-            # Delete from S3
-            # S3 delete is called first => if this fails the DB delete won't take place
+            # S3 delete is called first, if this fails for any other reason that the file not being
+            # found, then the DB delete won't take place, and an exception raised.
             dataset_key = self._get_s3_key(record.id, record.filename)
             dataset_path = f"s3://{ Path(settings.S3_BUCKET) / dataset_key }"
             self.s3_filesystem.rm(dataset_path, recursive=True)
+        except FileNotFoundError as e:
+            # File not found errors are allowed, but we perform clean-up in this situation.
+            logger.error(f"Dataset ID: {dataset_id} was present in the DB but not found on S3... "
```

Comment:
Being non-breaking, I'd keep this as a logger.info or a logger.warning

Reply:
https://github.com/mozilla-ai/lumigator/pull/332/commits/91169a6e368869b617dc5cde14a4ab61b1b47626